cdbot/cogs/admin.py

```diff
@@ -89,6 +92,32 @@ async def on_member_join(self, member: Member):
             # assign placeholder nickname
             await member.edit(nick=PLACEHOLDER_NICKNAME)
 
+    @Cog.listener()
+    async def on_message(self, message):
+        def check(m):
+            return m.author == message.author and (datetime.datetime.utcnow() - m.created_at).seconds < 4
+
+        def who(m):
+            return m.author == message.author
+
+        if not message.author.bot:
+            # Checks if message contains banned word.
+            if any([word in message.content for word in BANNED_WORDS]):
+                await message.delete()
+                await message.channel.send(f"{message.author.mention} Watch your language!", delete_after=5)
+            # Checks if message contains banned domain.
+            if any([word in message.content for word in BANNED_DOMAINS]):
+                await message.delete()
+                await message.channel.send(f"{message.author.mention} No invite links!", delete_after=5)
+            # Checks if message contains mass mention.
+            if len(message.mentions) > 8:
+                await message.delete()
+                await message.channel.send(f"{message.author.mention} Don't spam mentions!", delete_after=5)
+            # Checks if message was sent too quickly.
+            if len(list(filter(lambda m: check(m), self.client.cached_messages))) >= 4:
```

Comment:
`self.client.cached_messages` is an array of 1000 messages. Ignoring the fact that the lambda is redundant, this will call `check` 1000 times for every message sent to the bot. Rather than checking every message retroactively, it would be better to track timestamps of sent messages as a dictionary of dequeues, each limited to 5 items, indexed by user ID. Each dequeue would contain the timestamp of messages as the bot receives them. Checking for spam is then a case of pulling up the corresponding dequeue for a given user, and checking if the oldest item in the dequeue is younger than n seconds. At this moment, spamming is more likely to DoS the bot than be deleted.